src/lib.rs

```diff
@@ -568,15 +568,33 @@ pub unsafe trait AsBytes {
     /// `as_bytes` provides access to the bytes of this value as an immutable
     /// byte slice.
     fn as_bytes(&self) -> &[u8] {
-        // TODO(#61): Add a "SAFETY" comment and remove this `allow`.
-        #[allow(clippy::undocumented_unsafe_blocks)]
-        unsafe {
-            // Note that this method does not have a `Self: Sized` bound;
-            // `size_of_val` works for unsized values too.
-            let len = mem::size_of_val(self);
-            let slf: *const Self = self;
-            slice::from_raw_parts(slf.cast::<u8>(), len)
-        }
+        // Note that this method does not have a `Self: Sized` bound;
+        // `size_of_val` works for unsized values too.
+        let len = mem::size_of_val(self);
+        let slf: *const Self = self;
+
+        // SAFETY:
+        // - `slf.cast::<u8>()` is valid for reads for `len *
+        //   mem::size_of::<u8>()` many bytes because...
+        //   - `slf` is the same pointer as `self`, and `self` is a reference
+        //     which points to an object whose size is `len`. Thus...
+        //     - The entire region of `len` bytes starting at `slf` is contained
+        //       within a single allocation.
+        //     - `slf` is non-null.
+        //   - `slf` is trivially aligned to `align_of::<u8>() == 1`.
+        // - `Self: AsBytes` ensures that all of the bytes of `slf` are
+        //   initialized.
+        // - Since `slf` is derived from `self`, and `self` is an immutable
+        //   reference, the only other references to this memory region that
+        //   could exist are other immutable references, and those don't allow
+        //   mutation.
```

Comment:
This argument is technically not sufficient, as a `Cell<usize>` or other type with interior mutability could mutate the bytes after getting the byte slice. This case is covered by the safety requirements of `AsBytes` which state that implementors guarantee that it is sound to treat the type as a byte slice of `size_of::<T>()` len.

I think an appropriate change here would be:
- More explicitly ban interior mutability for `AsBytes` types.
- Call out that `AsBytes` allows us to treat this as a byte slice.

Same thing below for `as_bytes_mut`.

Reply:
Yeah, we have that tracked: https://github.com/google/zerocopy/issues/8

I was hoping that the following `TODO` would capture that since, tbh, `AsBytes` is basically entirely unsound right now since the safety requirements don't mention interior mutability.

As for `as_bytes_mut`, my understanding is that interior mutability isn't relevant since `&mut self` ensures that there are no other references full stop. Does that sound right to you?

Reply:
Yep, that sounds right to me. If we have it tracked then LGTM